src/lib/pubkey/curve448/curve448_utils/curve448_gf.cpp

```diff
@@ -0,0 +1,399 @@
+                                0xffffffffffffffff};
+
+   std::array<uint64_t, 7> in_minus_p;
+   bool borrow = false;
+   for(int i = 0; i < 7; ++i) {
+      in_minus_p[i] = u64_sub_with_borrow(in[i], p[i], &borrow);
+   }
+   std::array<uint64_t, 7> out;
+   CT::Mask<uint64_t>::expand(borrow).select_n(out.data(), in.data(), in_minus_p.data(), 7);
+   return out;
+}
+
+}  // namespace
+
+Gf448Elem::Gf448Elem(std::span<const uint8_t, 56> x) {
+   load_le(m_x.data(), x.data(), m_x.size());
+}
+
+Gf448Elem::Gf448Elem(uint64_t least_sig_word) {
+   clear_mem(m_x);
+   m_x[0] = least_sig_word;
+}
+
+void Gf448Elem::to_bytes(std::span<uint8_t, 56> out) const {
+   const auto canonical_form = to_canonical(m_x);
+   copy_out_le(out.data(), out.size(), canonical_form.data());
+}
+
+std::array<uint8_t, 56> Gf448Elem::to_bytes() const {
+   std::array<uint8_t, 56> bytes;
+   to_bytes(bytes);
+   return bytes;
+}
+
+void Gf448Elem::ct_cond_swap(bool b, Gf448Elem& other) {
+   for(size_t i = 0; i < 7; ++i) {
+      CT::conditional_swap(b, m_x[i], other.m_x[i]);
+   }
+}
+
+void Gf448Elem::ct_cond_assign(bool b, const Gf448Elem& other) {
+   CT::conditional_assign_mem(static_cast<uint64_t>(b), m_x.data(), other.m_x.data(), 7);
+}
+
+Gf448Elem Gf448Elem::operator+(const Gf448Elem& other) const {
+   Gf448Elem res(0);
+   gf_add(res.m_x, m_x, other.m_x);
+   return res;
+}
+
+Gf448Elem Gf448Elem::operator-(const Gf448Elem& other) const {
+   Gf448Elem res(0);
+   gf_sub(res.m_x, m_x, other.m_x);
+   return res;
+}
+
+Gf448Elem Gf448Elem::operator-() const {
+   Gf448Elem res(0);
+   gf_sub(res.m_x, res.m_x, m_x);
+   return res;
+}
+
+Gf448Elem Gf448Elem::operator*(const Gf448Elem& other) const {
+   Gf448Elem res(0);
+   gf_mul(res.m_x, m_x, other.m_x);
+   return res;
+}
+
+Gf448Elem Gf448Elem::square() const {
+   Gf448Elem res(0);
+   gf_square(res.m_x, m_x);
+   return res;
+}
+
+Gf448Elem Gf448Elem::operator/(const Gf448Elem& other) const {
+   Gf448Elem res(0);
+   gf_inv(res.m_x, other.m_x);
+   gf_mul(res.m_x, m_x, res.m_x);
+   return res;
+}
+
+Gf448Elem Gf448Elem::root() const {
+   Gf448Elem res(1);
+
+   // (P-3)/4 is an 445 bit integer with one zero bits at 222. All others are one.
+   for(int16_t t = 445; t >= 0; --t) {
+      gf_square(res.m_x, res.m_x);
+      if(t != 222) {
+         gf_mul(res.m_x, res.m_x, m_x);
+      }
+   }
+
+   return res;
+}
+
+bool Gf448Elem::operator==(const Gf448Elem& other) const {
+   const auto canonical_form_this = to_canonical(m_x);
+   const auto canonical_form_other = to_canonical(other.m_x);
+   return CT::is_equal(canonical_form_this.data(), canonical_form_other.data(), 7).as_bool();
+}
+
+bool Gf448Elem::is_zero() const {
+   const auto canonical_form = to_canonical(m_x);
+
+   return CT::all_zeros(canonical_form.data(), 7).as_bool();
+}
+
+bool Gf448Elem::is_odd() const {
+   const auto canonical_form = to_canonical(m_x);
+   return canonical_form[0] & 1;
```

Comment:
Pedantics might say one should explicitly create a boolean result:

```suggestion
   return canonical_form[0] & 1 != 0;
```